### PR TITLE
Use upstream hashicorp/raft

### PR DESCRIPTION
**Pull request by SimonRichardson**
The juju/raft fork had got way, way behind. The changes we wanted to
upstream have been upstreamed so it makes sense to move back to the
original.

## QA steps

```sh
$ juju bootstrap lxd test
$ juju enable-ha
$ juju remove-machine -m controller 0 # or which ever is leader
$ juju debug-log -m controller
$ juju model-config -m controller logging-config="<root>=INFO;juju.worker.lease.raft=TRACE;juju.core.raftlease=TRACE;juju.worker.globalclockupdater.raft=TRACE"
```

Ensure that there are no errors.

### CODE REVIEW

**Review by manadart — APPROVED** [review]
Nice.

File: worker/raft/worker_test.go [hunk 1]
```diff
@@ -74,7 +73,7 @@ func (s *WorkerValidationSuite) TestValidateErrors(c *gc.C) {
 		"empty LocalID not valid",
 	}, {
 		func(cfg *raft.Config) { cfg.HeartbeatTimeout = time.Millisecond },
-		"validating raft config: Heartbeat timeout is too low",
+		"validating raft config: HeartbeatTimeout is too low",
```
Comment (SimonRichardson) [inline worker/raft/worker_test.go:76]: They changed this upstream.

File: worker/raft/rafttest/fixtures.go [hunk 2]
```diff
@@ -79,9 +78,6 @@ func (s *RaftFixture) DefaultConfig(id raft.ServerID) *raft.Config {
 	raftConfig.HeartbeatTimeout = 100 * time.Millisecond
 	raftConfig.ElectionTimeout = raftConfig.HeartbeatTimeout
 	raftConfig.LeaderLeaseTimeout = raftConfig.HeartbeatTimeout
-	raftConfig.Logger = log.New(&raftutil.LoggoWriter{
-		loggo.GetLogger("juju.worker.raft.raftclusterer_test_" + string(id)),
-		loggo.DEBUG,
-	}, "", 0)
+	raftConfig.Logger = raftutil.NewHCLLogger("test", loggo.GetLogger("juju.worker.raft.raftclusterer_test"+string(id)))
```
Comment (jameinel) [inline worker/raft/rafttest/fixtures.go:81]: Interesting that loggo.Logger is API compatible with HCL Logger

File: worker/raft/raftutil/logging.go [hunk 3]
```diff
@@ -16,10 +21,25 @@ type Logger interface {
...
+	logWriter := &LoggoWriter{
+		Logger: logger,
+		Level:  loggo.DEBUG,
+		Prefix: fmt.Sprintf("[%s] ", name),
+	}
+
+	return hclog.New(&hclog.LoggerOptions{
+		Name:   name,
+		Output: logWriter,
+		Level:  hclog.DefaultLevel,
```
Comment (jameinel) [inline worker/raft/raftutil/logging.go:43]: Looking at the HCL codebase, it seems they use a log level, which not terribly surprising actually overlaps (I was concerned that there might be extra levels, but only Critical which we don't make active use of).

**Review by jameinel — APPROVED** [review]

**Comment by SimonRichardson** [comment]
$$merge$$

**Pull request merged**